rust_snuba/rust_arroyo/src/processing/dlq.rs

```diff
@@ -121,6 +126,84 @@ impl<TPayload> DlqPolicy<TPayload> {
     }
 }
 
+// Wraps the DLQ policy and keeps track of messages pending produce/commit.
+type Futures<TPayload> = VecDeque<(u64, JoinHandle<BrokerMessage<TPayload>>)>;
+
+pub(crate) struct DlqPolicyWrapper<TPayload> {
+    dlq_policy: Option<DlqPolicy<TPayload>>,
+    runtime: Handle,
+    // This is a per-partition max
+    max_pending_futures: usize,
+    futures: BTreeMap<Partition, Futures<TPayload>>,
+}
+
+impl<TPayload: Clone + Send + Sync + 'static> DlqPolicyWrapper<TPayload> {
+    pub fn new(dlq_policy: Option<DlqPolicy<TPayload>>) -> Self {
+        let concurrency_config = ConcurrencyConfig::new(10);
+        DlqPolicyWrapper {
+            dlq_policy,
+            runtime: concurrency_config.handle(),
+            max_pending_futures: 1000,
+            futures: BTreeMap::new(),
+        }
+    }
+
+    // Removes all completed futures, then appends a future with message to be produced
+    // to the queue. Blocks if there are too many pending futures until some are done.
+    pub fn produce(&mut self, message: BrokerMessage<TPayload>) {
+        for (_p, values) in self.futures.iter_mut() {
+            while !values.is_empty() {
+                let len = values.len();
+                let (_, future) = &mut values[0];
+                if future.is_finished() {
+                    values.pop_front();
+                } else if len >= self.max_pending_futures {
+                    let res = self.runtime.block_on(future);
+                    if let Err(err) = res {
+                        tracing::error!("Error producing to DLQ: {}", err);
+                    }
+                    values.pop_front();
+                } else {
+                    break;
+                }
+            }
+        }
+
+        if let Some(dlq_policy) = &self.dlq_policy {
+            let task = dlq_policy.producer.produce(message.clone());
+            let handle = self.runtime.spawn(task);
+
+            self.futures
+                .entry(message.partition)
+                .or_default()
+                .push_back((message.offset, handle));
+        }
+    }
+
+    // Blocks until all messages up to the committable have been produced so
+    // they are safe to commit.
+    pub fn flush(&mut self, committable: HashMap<Partition, u64>) {
+        for (p, values) in self.futures.iter_mut() {
+            while !values.is_empty() {
+                let (offset, future) = &mut values[0];
+                if let Some(committable_offset) = committable.get(p) {
```

Comment:
I would actually turn this around, and iterate over `committable` outside, and then pick the inner `self.futures` to join. As the futures are ordered by offset, you can then break out of that loop early.

Reply:
updated